core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsIT.java

```diff
@@ -1086,14 +1085,49 @@ public void testIndicesDeleted() throws Exception {
         assertAcked(prepareCreate("test"));
         ensureYellow();
 
-        String masterNode1 = internalCluster().getMasterName();
+        final String masterNode1 = internalCluster().getMasterName();
         NetworkPartition networkPartition = new NetworkUnresponsivePartition(masterNode1, dataNode.get(), getRandom());
         internalCluster().setDisruptionScheme(networkPartition);
         networkPartition.startDisrupting();
         internalCluster().client(masterNode1).admin().indices().prepareDelete("test").setTimeout("1s").get();
+        // Don't restart the master node until we know the index deletion has taken effect on master.
+        // This overcomes a current known issue where a delete can return before cluster state changes
+        // (in this case, index deletion) take effect.  We want to ensure that the cluster state changes
+        // are applied to the master node before it is restarted.
+        int numTries = 0;
+        final int maxNumTries = 20; // enforces that we don't end up in an infinite loop should something go wrong
+        final long sleepTime = 200L;
+        while (numTries < maxNumTries) {
+            final ClusterState currState = internalCluster().clusterService(masterNode1).state();
+            if (currState.metaData().hasIndex("test") == false && currState.status() == ClusterState.ClusterStateStatus.APPLIED) {
+                break;
+            }
+            try {
+                Thread.sleep(sleepTime);
```

Comment:
this is usually a bad sign. We should use sleep anywhere. Sometimes it's needed but we try give all the utilities to make sure no one used it explicitly. In this case we have assert busy:

```
        assertBusy(() -> {
            final ClusterState currState = internalCluster().clusterService(masterNode1).state();
            assertTrue("index not deleted", currState.metaData().hasIndex("test") == false && currState.status() == ClusterState.ClusterStateStatus.APPLIED);
        });
```